src/core/utils/otns.cpp

```diff
@@ -0,0 +1,136 @@
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements OTNS utilities.
+ *
+ */
+
+#include "otns.hpp"
+
+#if (OPENTHREAD_MTD || OPENTHREAD_FTD) && OPENTHREAD_CONFIG_OTNS_ENABLE
+
+#include "common/debug.hpp"
+#include "common/locator-getters.hpp"
+
+namespace ot {
+namespace Utils {
+
+const int kMaxStatusStringLength = 128;
+
+void Otns::EmitShortAddress(uint16_t aShortAddress)
+{
+    EmitStatus("rloc16=%d", aShortAddress);
+}
+
+void Otns::EmitExtendedAddress(const Mac::ExtAddress &aExtAddress)
+{
+    Mac::ExtAddress revExtAddress;
+    revExtAddress.Set(aExtAddress.m8, Mac::ExtAddress::kReverseByteOrder);
+    EmitStatus("extaddr=%s", revExtAddress.ToString().AsCString());
+}
+
+void Otns::EmitPingRequest(const Ip6::Address &aPeerAddress,
+                           uint16_t            aPingLength,
+                           uint32_t            aTimestamp,
+                           uint8_t             aHopLimit)
+{
+    OT_UNUSED_VARIABLE(aHopLimit);
+    EmitStatus("ping_request=%s,%d,%lu", aPeerAddress.ToString().AsCString(), aPingLength, aTimestamp);
+}
+
+void Otns::EmitPingReply(const Ip6::Address &aPeerAddress, uint16_t aPingLength, uint32_t aTimestamp, uint8_t aHopLimit)
+{
+    EmitStatus("ping_reply=%s,%u,%lu,%d", aPeerAddress.ToString().AsCString(), aPingLength, aTimestamp, aHopLimit);
+}
+
+void Otns::EmitStatus(const char *aFmt, ...)
+{
+    static char statusStr[kMaxStatusStringLength + 1];
+    int         n;
+
+    va_list ap;
+    va_start(ap, aFmt);
+
+    n = vsnprintf(statusStr, sizeof(statusStr), aFmt, ap);
+    OT_ASSERT(n >= 0);
+
+    va_end(ap);
+
+    otPlatOtnsStatus(statusStr);
+}
+
+void Otns::HandleStateChanged(Notifier::Callback &aCallback, otChangedFlags aFlags)
+{
+    aCallback.GetOwner<Otns>().HandleStateChanged(aFlags);
+}
+
+void Otns::HandleStateChanged(otChangedFlags aFlags)
+{
+    if ((aFlags & OT_CHANGED_THREAD_ROLE) != 0)
+    {
+        EmitStatus("role=%d", Get<Mle::Mle>().GetRole());
+    }
+
+    if ((aFlags & OT_CHANGED_THREAD_PARTITION_ID) != 0)
+    {
+        EmitStatus("parid=%x", Get<Mle::Mle>().GetLeaderData().GetPartitionId());
+    }
+
+    if ((aFlags & OT_CHANGED_JOINER_STATE) != 0)
+    {
+        EmitStatus("joiner_state=%d", Get<MeshCoP::Joiner>().GetState());
+    }
+}
+
+void Otns::EmitNeighborChange(otNeighborTableEvent aEvent, Neighbor &aNeighbor)
+{
+    const char *extAddr = aNeighbor.GetExtAddress().ToString().AsCString();
```

Comment:
This won't be correct.

Note that `ToString()` returns a temporary `String` object (think of it as returning an struct containing an array of chars), and `AsCString()` gets the pointer to the array.

A temp object stays valid for the whole enclosing statement/expression (in this case the whole statement is the `const char *extAddr = ...`). After this statement the temp object is destroyed/freed but we will keep a pointer to a freed char buffer in `extAddr`. From c++ [reference](https://en.cppreference.com/w/cpp/language/lifetime#Temporary_object_lifetime): "All temporary objects are destroyed as the last step in evaluating the full-expression that (lexically) contains the point where they were created"